cpp/src/endpoint.cpp

```diff
@@ -205,7 +219,27 @@ void Endpoint::removeInflightRequest(const Request* const request)
   _inflightRequests->remove(request);
 }
 
-size_t Endpoint::cancelInflightRequests() { return _inflightRequests->cancelAll(); }
+size_t Endpoint::cancelInflightRequests()
+{
+  auto worker     = ::ucxx::getWorker(this->_parent);
+  size_t canceled = 0;
+
+  if (worker->isProgressThreadRunning()) {
+    utils::CallbackNotifier callbackNotifierPre{false};
+    worker->registerGenericPre([this, &callbackNotifierPre, &canceled]() {
+      canceled = _inflightRequests->cancelAll();
+      callbackNotifierPre.store(true);
+    });
+    callbackNotifierPre.wait([](auto flag) { return flag; });
+    utils::CallbackNotifier callbackNotifierPost{false};
+    worker->registerGenericPost([&callbackNotifierPost]() { callbackNotifierPost.store(true); });
+    callbackNotifierPost.wait([](auto flag) { return flag; });
```

Comment:
The barrier is just to ensure cancelation was indeed progressed, otherwise there's always a chance we destroy internal resources before that completes, which could cause undefined behavior.